<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Extract putative causal variants within a candidate gene from a tabix-indexed snpEff annotated VCF file. — extract_variant • panGenomeBreedr</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Extract putative causal variants within a candidate gene from a tabix-indexed snpEff annotated VCF file. — extract_variant"><meta name="description" content="Extract putative causal variants within a candidate gene from a tabix-indexed
snpEff annotated VCF file."><meta property="og:description" content="Extract putative causal variants within a candidate gene from a tabix-indexed
snpEff annotated VCF file."><meta property="og:image" content="https://awkena.github.io/panGenomeBreedr/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">panGenomeBreedr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/awkena/panGenomeBreedr" aria-label="GitHub repository"><span class="fa fab fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Extract putative causal variants within a candidate gene from a tabix-indexed snpEff annotated VCF file.</h1>
      <small class="dont-index">Source: <a href="https://github.com/awkena/panGenomeBreedr/blob/HEAD/R/variant_annotation.R" class="external-link"><code>R/variant_annotation.R</code></a></small>
      <div class="d-none name"><code>extract_variant.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Extract putative causal variants within a candidate gene from a tabix-indexed
snpEff annotated VCF file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">extract_variant</span><span class="op">(</span></span>
<span>  <span class="va">cand_gene_id</span>,</span>
<span>  <span class="va">gff_path</span>,</span>
<span>  <span class="va">vcf_dir</span>,</span>
<span>  <span class="va">vcf_file</span>,</span>
<span>  output_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  outfile_suffix <span class="op">=</span> <span class="st">"variants"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-cand-gene-id">cand_gene_id<a class="anchor" aria-label="anchor" href="#arg-cand-gene-id"></a></dt>
<dd><p>A character value specifying the candidate gene ID.</p></dd>


<dt id="arg-gff-path">gff_path<a class="anchor" aria-label="anchor" href="#arg-gff-path"></a></dt>
<dd><p>A character value indicating the path to the GFF file,
including the complete file name.</p></dd>


<dt id="arg-vcf-dir">vcf_dir<a class="anchor" aria-label="anchor" href="#arg-vcf-dir"></a></dt>
<dd><p>A character value indicating the path to directory containing
snpEff annotated VCF files.</p></dd>


<dt id="arg-vcf-file">vcf_file<a class="anchor" aria-label="anchor" href="#arg-vcf-file"></a></dt>
<dd><p>A character value indicating the file name for snpEff
annotated VCF file including the .vcf.gz extension.</p></dd>


<dt id="arg-output-path">output_path<a class="anchor" aria-label="anchor" href="#arg-output-path"></a></dt>
<dd><p>A character value indicating the path to directory for
saving extracted variants.</p></dd>


<dt id="arg-outfile-suffix">outfile_suffix<a class="anchor" aria-label="anchor" href="#arg-outfile-suffix"></a></dt>
<dd><p>A character value indicating the file name to be used for
saving extracted variants.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A VCF file containing extracted variants for <code>cand_gene_id</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This wrapper function operates on tabix-indexed snpEff annotated VCF files.
However, if a tabix-indexed VCF file is not available, it can create one from
the inputted VCF file.</p>
<p>The file names of snpEff annotated VCF files are expected to consist of three
components: a common prefix, chromosome tag and a common suffix.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># example code</span></span></span>
<span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://awkena.github.io/panGenomeBreedr">panGenomeBreedr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Work from the tempdir</span></span></span>
<span class="r-in"><span><span class="va">vcf_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Google drive link to gff3 file</span></span></span>
<span class="r-in"><span><span class="va">flink1</span> <span class="op">&lt;-</span> <span class="st">"https://drive.google.com/file/d/1XjYyJ2JLywbbniIU6oUIIxAmEBKfmHpz/view?usp=sharing"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Download gff3 file to tempdir()</span></span></span>
<span class="r-in"><span><span class="va">gff3</span> <span class="op">&lt;-</span> <span class="fu"><a href="folder_download_gd.html">folder_download_gd</a></span><span class="op">(</span>drive_link <span class="op">=</span> <span class="va">flink1</span>,</span></span>
<span class="r-in"><span>                           output_path <span class="op">=</span> <span class="va">vcf_dir</span>,</span></span>
<span class="r-in"><span>                           is.folder <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> File downloaded:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> • <span style="color: #00BBBB;">Sbicolor_730_v5.1.gene.gff3</span> <span style="color: #555555;">&lt;id: 1XjYyJ2JLywbbniIU6oUIIxAmEBKfmHpz&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saved locally as:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> • <a href="file:///var/folders/tf/jr_ht05123v5x1196mv9z5zr0000gn/T//Rtmpc3Cm8N/Sbicolor_730_v5.1.gene.gff3"><span style="color: #0000BB;">/var/folders/tf/jr_ht05123v5x1196mv9z5zr0000gn/T//Rtmpc3Cm8N/Sbicolor_730_v5.1.gene.gff3</span></a></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Google drive link to indel snpEff annotated vcf file on Chr05</span></span></span>
<span class="r-in"><span><span class="va">flink2</span> <span class="op">&lt;-</span> <span class="st">"https://drive.google.com/file/d/1LiOeDsfIwbsCuHbw9rCJ1FLOZqICTrfs/view?usp=sharing"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Download indel snpEff annotated vcf file to tempdir()</span></span></span>
<span class="r-in"><span><span class="va">vcf_file_indel</span> <span class="op">&lt;-</span> <span class="fu"><a href="folder_download_gd.html">folder_download_gd</a></span><span class="op">(</span>drive_link <span class="op">=</span> <span class="va">flink2</span>,</span></span>
<span class="r-in"><span>                                     output_path <span class="op">=</span> <span class="va">vcf_dir</span>,</span></span>
<span class="r-in"><span>                                     is.folder <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> File downloaded:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> • <span style="color: #00BBBB;">Sorghum_d8.noduplicates.Chr05.indel._markernamesadded_imputed_snpeff.vcf.gz</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #555555;">&lt;id: 1LiOeDsfIwbsCuHbw9rCJ1FLOZqICTrfs&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saved locally as:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> • <a href="file:///var/folders/tf/jr_ht05123v5x1196mv9z5zr0000gn/T//Rtmpc3Cm8N/Sorghum_d8.noduplicates.Chr05.indel._markernamesadded_imputed_snpeff.vcf.gz"><span style="color: #0000BB;">/var/folders/tf/jr_ht05123v5x1196mv9z5zr0000gn/T//Rtmpc3Cm8N/Sorghum_d8.noduplicates.Chr05.indel._markernamesadded_imputed_snpeff.vcf.gz</span></a></span>
<span class="r-in"><span><span class="co"># View downloaded files in tempdir</span></span></span>
<span class="r-in"><span><span class="co"># list.files(vcf_dir)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># InDel variant extraction for lgs1 (Sobic.005G213600)</span></span></span>
<span class="r-in"><span><span class="fu">extract_variant</span><span class="op">(</span>cand_gene_id <span class="op">=</span> <span class="st">'Sobic.005G213600'</span>,</span></span>
<span class="r-in"><span>                gff_path <span class="op">=</span> <span class="va">gff3</span>,</span></span>
<span class="r-in"><span>                vcf_dir <span class="op">=</span> <span class="va">vcf_dir</span>,</span></span>
<span class="r-in"><span>                vcf_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">vcf_file_indel</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                output_path <span class="op">=</span> <span class="va">vcf_dir</span>,</span></span>
<span class="r-in"><span>                outfile_suffix <span class="op">=</span> <span class="st">'lgs_variants_indel'</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Registered S3 methods overwritten by 'Seqinfo':</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   method                from        </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   as.data.frame.Seqinfo GenomeInfoDb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   merge.Seqinfo         GenomeInfoDb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   summary.Seqinfo       GenomeInfoDb</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>replacing previous import ‘GenomeInfoDb::seqinfo&lt;-’ by ‘GenomicRanges::seqinfo&lt;-’ when loading ‘rtracklayer’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>replacing previous import ‘GenomeInfoDb::seqinfo’ by ‘GenomicRanges::seqinfo’ when loading ‘rtracklayer’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>replacing previous import ‘GenomeInfoDb::seqnames’ by ‘GenomicRanges::seqnames’ when loading ‘rtracklayer’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>replacing previous import ‘GenomeInfoDb::seqnames&lt;-’ by ‘GenomicRanges::seqnames&lt;-’ when loading ‘rtracklayer’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>replacing previous import ‘GenomeInfoDb::seqinfo&lt;-’ by ‘GenomicRanges::seqinfo&lt;-’ when loading ‘GenomicAlignments’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>replacing previous import ‘GenomeInfoDb::seqinfo’ by ‘GenomicRanges::seqinfo’ when loading ‘GenomicAlignments’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>replacing previous import ‘GenomeInfoDb::seqnames’ by ‘GenomicRanges::seqnames’ when loading ‘GenomicAlignments’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>replacing previous import ‘GenomeInfoDb::seqnames&lt;-’ by ‘GenomicRanges::seqnames&lt;-’ when loading ‘GenomicAlignments’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>replacing previous import ‘SummarizedExperiment::seqinfo’ by ‘Biostrings::seqinfo’ when loading ‘GenomicAlignments’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>replacing previous import ‘SummarizedExperiment::seqinfo&lt;-’ by ‘Biostrings::seqinfo&lt;-’ when loading ‘GenomicAlignments’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>multiple methods tables found for ‘seqinfo’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>multiple methods tables found for ‘seqinfo&lt;-’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>multiple methods tables found for ‘seqnames’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>multiple methods tables found for ‘seqnames&lt;-’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>multiple methods tables found for ‘seqinfo’</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>multiple methods tables found for ‘seqinfo&lt;-’</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Clean tempdir after variant extraction</span></span></span>
<span class="r-in"><span><span class="co"># unlink(vcf_dir, recursive = TRUE)</span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alexander Wireko Kena, Israel Tawiah Tetteh, Clara Cruet Burgos, Geoffrey Preston Morris.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

